### ISSUE TITLE
Typechecker rejects well-typed first-order polymorphic records on aliased degenerate types

### ISSUE THREAD

**Issue by dlesbre** [ocaml/ocaml#13292]
The typechecker improperly rejects code when expecting a record field of type `'a. ('a, 'b) foo` when `foo` does not depend on `'a`. Here is a minimal example (by discuss user octachron) :
```ocaml
module Alias(X: sig type ('a, 'b) snd end) = struct
  type ('a, 'b) alias = ('a, 'b) X.snd
end

module A = Alias(struct type ('a, 'b) snd = 'b end)
type 'b f = { f : 'a.  ('a, 'b) A.alias -> unit  }

let error (g: int -> unit) : int f = { f = g }
let also_error (g: int -> unit) : int f = { f = Obj.magic g }
```
This code is properly typed, but the typechecker rejects it:
```
8 | let error (g: int -> unit) : int f = { f = g }
                                               ^
Error: This field value has type int -> unit which is less general than
         'a. 'b -> unit
```
Adding an `Obj.magic` does not help either, the return value of Obj.magic still has an incorrect type:
```
9 | let also_error (g: int -> unit) : int f = { f = Obj.magic g }
                                                    ^^^^^^^^^^^
Error: This field value has type int -> unit which is less general than
         'a. 'b -> unit
```
The typecheck appears unable to expand the type alias and realise that `'a. ('a, int) A.alias` is just `'a. int`.

I've also made a [discuss post](https://discuss.ocaml.org/t/weird-behaviors-with-first-order-polymorphism/13783) about this issue, back when I wasn't conviced it was an error in OCaml. The replies helpfully clarified the situation and suggested a (somewhat) acceptable workaround: instead of using a direct alias `type ('a, 'b) snd = 'b`, use an wrapped (unboxed) type: `type ('a, 'b) snd = Snd of 'b [@@unboxed]`. This avoids the above error, but necessitates adding constructors/destructors around function calls.

This was confirmed on both OCaml 4.14.1 and 5.2.

**Comment by garrigue** [comment]
This problem actually comes from your use of a functor.
Namely, `update_level`, the function tracking the scope of type variables, uses the variance inferred for types to decide whether it should expand an abbreviation of not (expanding all abbreviations would have some adverse effects).
And this variance information is incorrect for the result of functors (namely, the variance is the one inferred for the functor itself, which assumes all abstract types in parameter to have only their declared variance. It is a long standing issue. The only solution I see would be to recompute the variances in the result of functors, maybe when we add them to the environment.
If you define `A` directly rather than through a functor, this solves the problem.
```ocaml
module A = struct type ('a,'b) alias = 'b end
```

**Comment by dlesbre** [comment]
Yes, but inlining the functor arguments comes with a lot of code duplication I'm afraid.

 For context I'm using this as a part of a [patricia-tree library](https://codex.top/api/patricia-tree/) which has both heterogeneous maps (`'b map` mapping `'k key` to `('k, 'b) value`) and homogeneous maps. The issue crops up when [trying to instantiate the heterogeneous map functors with homogeneous types](https://github.com/codex-semantics-library/patricia-tree/blob/4b08d346bb058de056f5a54b5a42ea020a141e74/src/functors.ml#L1117) (`'k key = key` and `(_, 'b) value = 'b`).

**Comment by Octachron** [comment]
For the functor variance issue, could we have variance variables at the functor level? My impression is that being able to declare that a functor is transporting (or negating) variances of the functor argument would be enough in many situations.

**Comment by gasche** [comment]
Variance variables sound like hard work. Is there a good reason not to recompute variances on functor applications? I wouldn't expect a noticeable performance cost coming from this, type definitions are small and variance computation is fast. Is it just that no one has been motivated to do the work?

**Comment by garrigue** [comment]
Indeed, this is mostly a problem of motivation.
I had a look at the code last week. The natural place to call the recomputation would be in `env.ml`,
but then you are in the middle of a complex data-structure, and it is not so easy to find out all you need to call the appropriate function of `typedecl_variance.ml`.
Should try to do something about it, but it looks like it might be more than an afternoon hack.
Also, while the cost should not be high, you still want to trigger it only on the result of functor applications, and do not break their laziness.

**Comment by lpw25** [comment]
> I wouldn't expect a noticeable performance cost coming from this, type definitions are small and variance computation is fast.

Module types in the results of functors can be huge. Simply traversing them looking for type definitions can be very expensive. We should be trying very hard to never traverse module types deeply, so I'm opposed to basically any change that adds more of this.

**Comment by lpw25** [comment]
(I think Jacques suggested approach is probably fine for this -- as he says you need to not break the laziness of the application -- I just wanted to push back on the idea that doing anything deeply over the result of a functor application is cheap).
